cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -13,206 +13,102 @@
+  (arg = 1 or arg = 2)
 }
 
-predicate underscoreMacro(Expr e) {
-  exists(MacroInvocation mi |
-    mi.getMacroName() = "_" and
-    mi.getExpr() = e
-  )
+predicate whitelisted(FunctionCall fc) {
+  exists(Function f, int arg | f = fc.getTarget() | whitelistFunction(f, arg))
 }
 
-/**
- * Holds if a value of character pointer type may flow _directly_ from `src` to
- * `dst`.
- */
-predicate stringFlowStep(Expr src, Expr dst) {
-  not underscoreMacro(dst)
-  and
-  stringType(dst.getType(), _)
-  and
-  (
-    src = dst.(VariableAccess).getTarget().getAnAssignedValue()
+predicate isNonConst(DataFlow::Node node) {
+  exists(Expr e | e = node.asExpr() |
+    exists(FunctionCall fc | fc = e.(FunctionCall) |
+      not whitelisted(fc) and not fc.getTarget().hasDefinition()
+    )
+    or
+    exists(Parameter p | p = e.(VariableAccess).getTarget().(Parameter) |
+      p.getFunction().getName() = "main" and p.getType() instanceof PointerType
+    )
     or
-    src = dst.(ConditionalExpr).getThen()
+    e instanceof CrementOperation
     or
-    src = dst.(ConditionalExpr).getElse()
+    e instanceof AddressOfExpr
     or
-    src = dst.(AssignExpr).getRValue()
+    e instanceof ReferenceToExpr
     or
-    src = dst.(CommaExpr).getRightOperand()
+    e instanceof AssignPointerAddExpr
     or
-    src = dst.(UnaryPlusExpr).getOperand()
+    e instanceof AssignPointerSubExpr
     or
-    stringArray(dst.(ArrayExpr).getArrayBase().(VariableAccess).getTarget(),
-                src.getParent())
+    e instanceof PointerArithmeticOperation
     or
-    // Follow a parameter to its arguments in all callers.
-    exists(Parameter p | p = dst.(VariableAccess).getTarget() |
-      src = p.getFunction().getACallToThisFunction().getArgument(p.getIndex())
-    )
+    e instanceof FieldAccess
     or
-    // Follow a call to a gettext function without looking at its body even if
-    // the body is known. This ensures that we report the error in the relevant
-    // location rather than inside the body of some `_` function.
-    exists(Function gettext, FunctionCall call, int idx |
-      gettextFunction(gettext, idx) and
-      dst = call and
-      gettext = call.getTarget()
-    | src = call.getArgument(idx)
-    )
+    e instanceof PointerDereferenceExpr
+    or
+    e instanceof AddressOfExpr
     or
-    // Follow a call to a non-gettext function through its return statements.
-    exists(Function f, ReturnStmt retStmt |
-      f = dst.(FunctionCall).getTarget() and
-      f = retStmt.getEnclosingFunction() and
-      not gettextFunction(f, _)
-    | src = retStmt.getExpr()
+    e instanceof ExprCall
+    or
+    e instanceof NewArrayExpr
+    or
+    e instanceof AssignExpr
+    or
+    exists(Variable v | v = e.(VariableAccess).getTarget() |
+      v.getType().(ArrayType).getBaseType() instanceof CharType and
+      exists(AssignExpr ae |
+        ae.getLValue().(ArrayExpr).getArrayBase().(VariableAccess).getTarget() = v
+      )
```

Comment:
A goal with this PR was to improve performance, and I suspect you can do that by making `isNonConst` smaller. It currently includes all sorts of expressions that have nothing to do with strings. The existing query had a `stringType` predicate that was used in various places to cut down the number of results.

Reply:
Which disjuncts do you believe can be removed?

Reply:
I mean the latter.

Reply:
As per our discussion in Ox, adding the `stringType` check does not improve performance.